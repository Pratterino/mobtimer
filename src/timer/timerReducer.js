import actions from '../actionTypes';
import { store } from './../store';
import { getParsedTimeRemaining } from '../helper/TimerHelper';
import { activeUserSelector } from './../user/userReducer';
import { speak } from './../helper/Speech';
import { closeNotification, showNotification } from '../NotificationManager';
import FirebaseManager from './../FirebaseManager';

let interval;
let speechTimeout;
let titleToggleInterval;

export const defaultTimerState = {
    active: false,
    sessionLength: 60 * 15,
    currentTime: 0,
    metadata: {
        todaysSessionLength: 0,
        todaysDate: new Date().getDate(),
    },
    leaderboard: {},
};

const timerIsDone = () => {
    // ALARM
    store.dispatch({
        type: actions.PLAY_FINISH_SOUND,
    });

    store.dispatch({
        type: actions.FINISH_TIMER,
    });

    // todo: thunk it =>
    store.dispatch({
        type: actions.NEXT_USER,
    });

    showNextUserNotification();
};

const showNextUserNotification = () => {
    const upNextUser = activeUserSelector(store.getState());
    showNotification(upNextUser);
    console.info('NOTIFICATION: ', upNextUser);
};

const stopTimerInterval = () => {
    clearInterval(interval);
    console.info('TIMER: CLEARED INTERVAL');
    return () => store.dispatch(updateState);
};

const updateState = (dispatch, getState) => {
    return FirebaseManager.updateState(getState());
};

const startTimerInterval = () => {
    clearTimeout(speechTimeout);
    clearInterval(titleToggleInterval);

    interval = setInterval(() => {
        let state = store.getState();
        if (state.timer.currentTime <= 0) {
            timerIsDone();
            store.dispatch({
                type: actions.RESET_TIMER,
            });
            return;
        }
        store.dispatch({
            type: actions.SECOND_DECREMENT_TIMER,
            user: activeUserSelector(state),
        });

        store.dispatch(updateState);
    }, 1000);
    console.info('TIMER: MOB SESSION STARTED!');
    closeNotification();
};

const timeoutToSpeech = (i = 0) => {
    if (i >= 3) {
        speak(`${i} minutes! I give up. I'll stop talking to you. Good bye, asshole`);
        showNextUserNotification();
        return clearTimeout(speechTimeout);
    } else {
        speechTimeout = setTimeout(() => {
            const state = store.getState();
            const activeUser = activeUserSelector(state);

            showNextUserNotification();
            speak(
                `It's ${activeUser.name}'s time! You've been idle for an entire ${
                    i >= 1 ? `${i + 1} minutes!` : 'minute!'
                }`,
            );
            return timeoutToSpeech(i + 1);
        }, 60 * 1000);
    }
    console.info(`TIMER: SPEECH COUNTDOWN 60s, iteration ${i + 1}`);
};

const toggleTitleOnFinish = () => {
    let interval = 0;
    titleToggleInterval = setInterval(() => {
        const state = store.getState();
        const activeUser = activeUserSelector(state);

        document.title = interval % 2 ? '⏰⏰⏰' : `${activeUser.name.toUpperCase()}`;
        interval++;
    }, 1000);
};

export default (state = defaultTimerState, action) => {
    switch (action.type) {
        case actions.SECOND_DECREMENT_TIMER:
            const seconds = state.currentTime - 1;
            document.title = `${getParsedTimeRemaining(seconds)}`; //TODO: useEffect()
            const leaderboardCurrentUserSeconds = state.leaderboard[action.user.name] || 0;
            return {
                ...state,
                currentTime: seconds,
                metadata: {
                    ...state.metadata,
                    todaysSessionLength: state.metadata.todaysSessionLength + 1,
                },
                leaderboard: {
                    ...state.leaderboard,
                    [action.user.name]: leaderboardCurrentUserSeconds + 1,
                },
            };

        case actions.RESET_TODAYS_SESSION_LENGTH:
            return {
                ...state,
                metadata: {
                    ...state.metadata,
                    todaysSessionLength: 0,
                },
                leaderboard: defaultTimerState.leaderboard,
            };

        case actions.START_TIMER:
            startTimerInterval();
            clearTimeout(speechTimeout);
            return {
                ...state,
                currentTime: state.sessionLength,
                active: true,
                metadata: {
                    ...state.metadata,
                    todaysDate: new Date().getDate(),
                },
            };

        case actions.FINISH_TIMER:
            // when a single timer cycle has completed
            stopTimerInterval();
            toggleTitleOnFinish();
            timeoutToSpeech();
            return {
                ...state,
                currentTime: state.sessionLength,
                active: false,
            };

        case actions.STOP_TIMER:
            stopTimerInterval();
            clearTimeout(speechTimeout);
            return {
                ...state,
                active: false,
            };

        case actions.PLAY_PAUSE_TIMER:
            if (state.currentTime === null) {
                // first time playing
            }
            if (state.active) {
                stopTimerInterval();
            } else {
                startTimerInterval();
            }

            clearTimeout(speechTimeout);

            return {
                ...state,
                currentTime: state.currentTime || state.sessionLength,
                active: !state.active,
                metadata: {
                    ...state.metadata,
                    todaysDate: new Date().getDate(),
                },
            };

        case actions.RESET_TIMER:
            stopTimerInterval();
            return {
                ...state,
                currentTime: state.sessionLength,
                active: false,
            };

        case actions.UPDATE_SESSION_LENGTH:
            stopTimerInterval();
            clearTimeout(speechTimeout);

            return {
                ...state,
                sessionLength: action.sessionLength,
            };

        default:
            return state;
    }
};
