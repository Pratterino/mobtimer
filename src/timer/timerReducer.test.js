import React from 'react';
import timerReducer, { defaultTimerState } from './timerReducer';
import actions from './../actionTypes';

describe('timerReducer', () => {
    const getDefaultState = props => ({
        ...defaultTimerState,
        ...{
            metadata: {
                ...defaultTimerState.metadata,
                ...{
                    todaysDate: 1,
                },
            },
        },
        ...props,
    });

    beforeEach(() => {
        jest.useFakeTimers();
    });

    //TODO: Test these "private" methods.
    describe('timerIsDone', () => {});
    describe('stopTimerInterval', () => {});
    describe('startTimerInterval', () => {});
    describe('timeoutToSpeech', () => {});
    describe('toggleTitleOnFinish', () => {});

    describe('SECOND_DECREMENT_TIMER', () => {
        let state;
        let action;

        beforeEach(() => {
            state = getDefaultState({
                currentTime: 1050,
            });
            state.metadata.todaysSessionLength = 41;

            action = {
                type: actions.SECOND_DECREMENT_TIMER,
                user: {
                    name: 'Ulla',
                },
            };
        });

        it('should decrement currentTime and increment todaysSessionLength', () => {
            let unknownState = timerReducer(state, { type: 'UNKNOWN' });
            let newState = timerReducer(state, action);

            expect(unknownState.currentTime).toBe(1050);
            expect(newState.currentTime).toBe(1049);
            expect(document.title).toBe('17:29');

            expect(unknownState.metadata.todaysSessionLength).toBe(41);
            expect(newState.metadata.todaysSessionLength).toBe(42);
        });
    });

    describe('RESET_TODAYS_SESSION_LENGTH', () => {
        let state;
        let action;

        beforeEach(() => {
            state = getDefaultState();
            state.metadata.todaysSessionLength = 500;

            action = {
                type: actions.RESET_TODAYS_SESSION_LENGTH,
            };
        });

        it('should reset todaysSessionLength to zero', () => {
            let unknownState = timerReducer(state, { type: 'UNKNOWN' });
            let newState = timerReducer(state, action);

            expect(unknownState.metadata.todaysSessionLength).toBe(500);
            expect(newState.metadata.todaysSessionLength).toBe(0);
        });
    });

    describe('START_TIMER', () => {
        let state;
        let action;

        beforeEach(() => {
            state = getDefaultState({
                active: true,
                currentTime: 62,
                sessionLength: 5000,
            });
            action = {
                type: actions.START_TIMER,
            };
        });

        it('should start interval and clear timeout', () => {
            let newState = timerReducer(state, action);

            expect(clearTimeout).toHaveBeenCalledTimes(2);
            expect(clearInterval).toHaveBeenCalledTimes(1);
            expect(newState.active).toBe(true);
            expect(newState.currentTime).toBe(state.sessionLength);
            expect(newState.metadata.todaysDate).toBe(new Date().getDate());
        });
    });

    describe('FINISH_TIMER', () => {
        let state;
        let action;

        beforeEach(() => {
            state = getDefaultState({
                active: true,
                currentTime: 62,
                sessionLength: 5000,
            });
            action = {
                type: actions.FINISH_TIMER,
            };
        });

        it('should call clearInterval', () => {
            let newState = timerReducer(state, action);

            expect(clearTimeout).not.toHaveBeenCalled();
            expect(clearInterval).toHaveBeenCalledTimes(1);
            expect(newState.active).toBe(false);
            expect(newState.currentTime).toBe(state.sessionLength);
        });

        it('should match snapshot', () => {
            let newState = timerReducer(state, action);
            expect(newState).toMatchSnapshot();
        });
    });

    describe('STOP_TIMER', () => {
        let state;
        let action;

        beforeEach(() => {
            state = getDefaultState({
                active: true,
            });
            action = {
                type: actions.STOP_TIMER,
            };
        });

        it('should call clearInterval and clearTimeout', () => {
            let newState = timerReducer(state, action);

            expect(clearTimeout).toHaveBeenCalledTimes(1);
            expect(clearInterval).toHaveBeenCalledTimes(1);

            expect(newState.active).toBe(false);
        });

        it('should match snapshot', () => {
            let newState = timerReducer(state, action);
            expect(newState).toMatchSnapshot();
        });
    });

    describe('PLAY_PAUSE_TIMER', () => {
        let state;
        let action;

        beforeEach(() => {
            action = {
                type: actions.PLAY_PAUSE_TIMER,
            };
        });

        it('should stop if active', () => {
            state = getDefaultState({
                currentTime: 43,
                active: true,
            });
            let newState = timerReducer(state, action);

            expect(clearTimeout).toHaveBeenCalledTimes(1);
            expect(clearInterval).toHaveBeenCalledTimes(1);

            expect(newState.active).toBe(false);
        });

        it('should start if not active', () => {
            state = getDefaultState({
                currentTime: 43,
                active: false,
            });
            let newState = timerReducer(state, action);

            expect(clearTimeout).toHaveBeenCalledTimes(2);
            expect(clearInterval).toHaveBeenCalledTimes(1);

            expect(newState.active).toBe(true);
        });

        describe('should match snapshots', function() {
            it('when active', () => {
                state = getDefaultState({
                    currentTime: 40,
                    active: true,
                });

                let newState = timerReducer(state, action);
                expect(newState.metadata.todaysDate).toBe(new Date().getDate());
                expect(newState.active).toBe(false);
            });

            it('when not active', () => {
                state = getDefaultState({
                    currentTime: 50,
                    active: false,
                });
                state.metadata.todaysDate = 12;

                let newState = timerReducer(state, action);
                expect(newState.metadata.todaysDate).toBe(new Date().getDate());
                expect(newState.active).toBe(true);
            });
        });
    });

    describe('RESET_TIMER', () => {
        let state;
        let action;

        beforeEach(() => {
            state = getDefaultState({
                active: true,
                currentTime: 60,
                sessionLength: 5000,
            });
            action = {
                type: actions.RESET_TIMER,
            };
        });

        it('should call clearInterval', () => {
            let newState = timerReducer(state, action);

            expect(clearTimeout).not.toHaveBeenCalled();
            expect(clearInterval).toHaveBeenCalledTimes(1);

            expect(newState.active).toBe(false);
            expect(newState.currentTime).toBe(5000);
        });

        it('should match snapshot', () => {
            let newState = timerReducer(state, action);
            expect(newState).toMatchSnapshot();
        });
    });

    describe('UPDATE_SESSION_LENGTH', () => {
        let state;

        beforeEach(() => {
            state = getDefaultState({
                sessionLength: 60,
            });
        });

        it('should call clearInterval and clearTimeout', () => {
            const action = {
                type: actions.UPDATE_SESSION_LENGTH,
                sessionLength: 120,
            };
            let newState = timerReducer(state, action);

            expect(clearTimeout).toHaveBeenCalledTimes(1);
            expect(clearInterval).toHaveBeenCalledTimes(1);

            expect(newState.sessionLength).toBe(120);
        });

        it('should update sessionLength and match snapshot', () => {
            const action = {
                type: actions.UPDATE_SESSION_LENGTH,
                sessionLength: 2000,
            };
            let newState = timerReducer(state, action);

            expect(newState.sessionLength).toBe(2000);
            expect(newState).toMatchSnapshot();
        });
    });

    describe('DEFAULT', () => {
        it('should return defaultState', function() {
            const action = {
                type: 'ACTION MAN',
            };
            let newState = timerReducer(undefined, action);
            expect(newState).toEqual(defaultTimerState);
        });
    });
});
