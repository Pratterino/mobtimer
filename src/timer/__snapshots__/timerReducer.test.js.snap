// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timerReducer FINISH_TIMER should match snapshot 1`] = `
Object {
  "active": false,
  "currentTime": 5000,
  "metadata": Object {
    "todaysDate": 1,
    "todaysSessionLength": 0,
  },
  "sessionLength": 5000,
}
`;

exports[`timerReducer RESET_TIMER should match snapshot 1`] = `
Object {
  "active": false,
  "currentTime": 5000,
  "metadata": Object {
    "todaysDate": 1,
    "todaysSessionLength": 0,
  },
  "sessionLength": 5000,
}
`;

exports[`timerReducer STOP_TIMER should match snapshot 1`] = `
Object {
  "active": false,
  "currentTime": 0,
  "metadata": Object {
    "todaysDate": 1,
    "todaysSessionLength": 0,
  },
  "sessionLength": 900,
}
`;

exports[`timerReducer UPDATE_SESSION_LENGTH should update sessionLength and match snapshot 1`] = `
Object {
  "active": false,
  "currentTime": 0,
  "metadata": Object {
    "todaysDate": 1,
    "todaysSessionLength": 0,
  },
  "sessionLength": 2000,
}
`;
